### ISSUE TITLE
Refactor RunConfig struct

### ISSUE THREAD

**Issue by davidcassany** [rancher/elemental-cli#33]
as per discussion in https://github.com/rancher-sandbox/elemental-cli/pull/32#discussion_r776967358 there is room to improve how we are processing and organizing run time configuration details. This can be actually our way to read the manifest in the various phases ( e.g. https://github.com/rancher-sandbox/elemental/pull/160 )

Ideas:
* Define a set of generic configs and pass it to methods instead of moving around the whole `RunConfig`
* Defile a config struct isolated from flags and config files (no marshall/unmarshall) for each subcommand (aka `InstallConfig` struct)

See also:
 https://github.com/rancher-sandbox/elemental/blob/f1228393e3f342c07e265850b7f4c06f0c0128d3/pkg/types/v1/config.go#L47

**Comment by mudler** [comment]
I would add to move validation outside from `cmd` and be strict with the config itself, this way it is detached from the cli framework being used, and can be used in different contexts (validate a config file, for example)

**Comment by davidcassany** [comment]
Here a proposal from my side (covers install/upgrade/reset commands):

 ```yaml
install:
  # target is the only mandatory
  target: /dev/sda

  # basic disk configs for formatting
  firmware: efi|bios
  partitionTable: gpt|msdos

  # partitions setup, any missing value is taken from defaults
  partitions:
    oem:
      label:
      size:
      fs:
      flags:
    recovery:
      ...
    state:
      ...
    persistent:
      ...

  # recovery image should be squashed or not (eventual future feature)
  squashfsRecovery: true|false

  # format disk or not
  noFormat: true|false

  # use this iso as installation media (overwrites 'image' and 'recoveryImage' sources with iso contents)
  iso:

  # main OS image
  image:
    label:
    size:
    fs:
    source:
      type: dir|channel|package|image|file
      uri:

  # these are recovery image details, uses the same as 'image' if unset
  recoveryImage:
    label:
    size:
    fs:
    source:
      type: dir|channel|package|image|file
      uri:
    
upgrade:
  # upgrade system to this OS Image
  image:
    label:
    size:
    fs:
    source:
      type: dir|channel|package|image|file
      uri:

  # upgrade recovery image to this setup
  recoveryImage:
    label:
    size:
    fs:
    source:
      type: dir|channel|package|image|file
      uri:
    
reset:
  # reset to this OS image
  image:
    label:
    size:
    fs:
    source:
      type: dir|channel|package|image|file
      uri:

  # format persistent partition or not
  formatPersistent: true|false

# channel repositories used for channel sources
repositories:
  - name:
    uri:
    priority:
    type:

# attempt a cosign validation
cosign: true|false
cosignKey: myKey

# attempt a verify process
verify: true|false

# fail on cloud-init hooks errors
strict: true|false

# reboot/power off when done
reboot: true|false
poweroff: true|false
 ```

I would suggest to map this manifest to a RunConfig struct. Note the most minimalistic setup in yaml form would be:

```yaml
install:
  target: /some/device
```

And a common and simple setup for derivatives:

```yaml
install:
  target: /some/device
  image:
    source:
      # type is guessed from uri if missing
      uri: registry.org/some/image:latest

upgrade:
  image:
    source:
      uri: registry.org/some/image:latest
```

@mudler @Itxaka @mjura @ldevulder any ideas or comments regarding this manifest setup to describe installation and upgrade commands (reset command is anecdotic)?

**Comment by ldevulder** [comment]
Sounds interesting. I didn't played a lot with upgrade for now but looks good too. Maybe Harvester's guys can have more thought on this?

**Comment by mudler** [comment]
Few inline thoughts. It looks very good overall, just few non-obvious things from my side

```yaml
 # use this iso as installation media (overwrites 'image' and 'recoveryImage' sources with iso contents)
 iso:

 # main OS image
 image:
   label:
   size:
   fs:
   source:
     type: dir|channel|package|image|file
     uri:

 # these are recovery image details, uses the same as 'image' if unset
 recoveryImage:
   label:
   size:
   fs:
   source:
     type: dir|channel|package|image|file
     uri:
```

Here seems a bit inconsistent as there is a `iso` section in the middle that have no other option behind to expose. Maybe we can have a `system`/`recovery` block stanza instead which condense a bit the URI passed by, just an example:

```yaml
install:
  system:
    source: iso | image
    uri: "type://<path>" # the uri here would have to be parsed to define what is the source from. In case of iso, can be a label which defaults to COS_LABEL, for example.
    ....
  recovery:
  ..
```

so you can do e.g.:

```yaml
install:
  system:
    source: "iso" # that should be default, so nothing to really set 
upgrade:
   system:
    source: "container://registry...."
```

Note I changed `image` here and go for something more generic like `system`as it might be confusing on how we refer to `image` in the different contexts, but this is just an example so anything else would fit.

**Comment by davidcassany** [comment]
I like the `system`/`recovery` stanzas. I also like the idea of having sources defined on per scheme basis, that could also allow us to compact `--docker-image` and `--directory` flags to `--system-source` and easily allow us to introduce a  `--recovery-source` flag. However I do not see clear the `source: "iso"`, as this is not exactly a source type as others and it can't be easily referenced on per scheme type basis. Probably we could have something like:

```yaml
install:
  system:
    source: docker://my.registry.org/my/image:latest
  recovery:
    source: docker://my.registry.org/my/recovery:latest
```
and 

```yaml
install:
  install-iso: http://domain.org/my.iso
```
and assume `install-iso` (or `install-media`?) can overwrite system and recovery default sources.

Note the `install-iso` maps to `--iso` flag and it is unset by default. The default installation source is a local directory (where the live ISOs are mounted).

**Comment by davidcassany** [comment]
Note in such a case the source is a custom URI type, we could accept something like `docker`, `file` and `channel-pkg`  as valid schemes and ignore anything else.

**Comment by mudler** [comment]
> I like the `system`/`recovery` stanzas. I also like the idea of having sources defined on per scheme basis, that could also allow us to compact `--docker-image` and `--directory` flags to `--system-source` and easily allow us to introduce a `--recovery-source` flag. However I do not see clear the `source: "iso"`, as this is not exactly a source type as others and it can't be easily referenced on per scheme type basis. Probably we could have something like:
> 
> ```yaml
> install:
>   system:
>     source: docker://my.registry.org/my/image:latest
>   recovery:
>     source: docker://my.registry.org/my/recovery:latest
> ```
> 
> and
> 
> ```yaml
> install:
>   install-iso: http://domain.org/my.iso
> ```
> 
> and assume `install-iso` (or `install-media`?) can overwrite system and recovery default sources.
> 
> Note the `install-iso` maps to `--iso` flag and it is unset by default. The default installation source is a local directory (where the live ISOs are mounted).

That sounds good here!

**Comment by Itxaka** [comment]
how would this work with "channel" packages? After all channel are just docker images but the version is auto selected to the latest by version number by luet, so you always get the latest version published. But we cannot use the latest tag as our published packages work a bit differently, like we use the category-name-version for the tag names....

**Comment by Itxaka** [comment]
something like `channel://system/grub2-efi-image` ?

**Comment by mudler** [comment]
> something like `channel://system/grub2-efi-image` ?

yup exactly, I thought at it as part of the `channel-pkg` mentioned above, to sum up we would have for example:
- `container` (for container images)
- `channel` (for channel upgrades)
- `file` (for local file paths)
- `http(s)` (for downloadable stuff)
- ...

and that would open up for further mechanisms on how we get the image in

**Comment by mudler** [comment]
@davidcassany  What I see missing here is the `live` packages related to iso bits like https://github.com/rancher-sandbox/cOS-toolkit/blob/master/manifest.yaml#L7, would that be a separate stanza? I don't see them being changed that much, so maybe just offering a way to override them in case of developing is needed.

@Itxaka for covering also derivatives with extra packages we could have a separate `extra` stanza or either having possibility to specify multiple sources, but that might be challenging. maybe we can just get away with having an `additional_sources` under the `iso` stanza.

**Comment by davidcassany** [comment]
I did some research regarding this card and so far I managed to find a way to determine a custom unmarshal method per type. So we can easily define a method in `v1.ImageSource` that actually translates the uri into an instance and this is used by viper's unmarshal. Note that in this particular case `http(s)` scheme is not useful to determine an image source as it could be a downloadable docker image, file or even luet package. But yes generally speaking we should consider https (and others) where it makes sense. The `net/url` turns to be really handy parse all this. It has two supported forms:

* `[scheme:][//[userinfo@]host][/]path[?query][#fragment]`
* `scheme:opaque[?query][#fragment]`

So our values after parsing could be expressed as `url.Opaque` or `url.Host + "/" + url.Path` and for the moment I would ignore any kind of user info, query and fragment.

**Comment by davidcassany** [comment]
> What I see missing here is the `live` packages related to iso bits, would that be a separate stanza? I don't see them being changed that much, so maybe just offering a way to override them in case of developing is needed.

With the current setup these belong to different files, even code wise we can share many things. So we have the runtime config which is part of the OS and the buildtime configs which are only used during build. So that the installation, upgrade and reset are stanzas to config those commands that expected to be executed inside a cOS system.

The current yaml exposed here is supposed to apply to RunConfig struct in elemental code and the `iso` and `disk` stanzas would be mapped to `BuildConfig` struct in elemental code. So different yaml files. To me this distinction makes sense and outlines a bit the use cases and assumptions we make in the code, but we could also place everything together.

**Comment by mudler** [comment]
By the way, we could also treat the iso as a special file case where we unpack the rootfs from differently. (like, we would unpack a "xz", a "gzip", an "iso") etc.

**Comment by davidcassany** [comment]
I created a new install command that only prints the loaded config with a completely new and parallel RunConfig, just as a play ground to experiment with different setups. I just created a PR of it so we can better discuss few details of it. Note this is not an implementation of anything, I just wanted to experiment with loading and unmarshaling more complex setups to see the limits and constraints we have.
